doc/REPLICATION.md

```diff
@@ -0,0 +1,138 @@
+
+Updates are pushed, or pulled during replication maintenance, if version numbers on both side of a connection does not match.
+Subscriptions are eventually consistent and converges to the highest version number. An author should not, but currently there
+is no protection against version number reuse.
+
+See #ReplicatorServer or (./api.proto) for gRPC API.
+
+Administrative
+
+Currently, the replication protocol requires it's own gRPC port. Which must be included with the servers libp2p address.
+
+When a server is asked to become active, it only checks one thing: is the author whitelisted?
+This also allows servers not listed to replicate the files but not counted towards replication factor.
+
+Security Details
+
+** This section is intended to help someone doing a security review of replication
+and should not be required reading to use the feature securely. **
+
+With security advice from:
+Protocol Interactions and the Chosen Protocol Attack
+(J. Kelsey, B. Schneier, and D. Wagner) 1997
+https://www.schneier.com/academic/archives/1997/01/protocol_interaction.html
+
+Protocol Buffers is designed for size and speed, but have the following properties that must be considered:
+
+	- Protocol Buffers are not self descripting.
+		Two protocols can easily share the same binary data structure,
+		but we must never sign or accept data for a different protocol.
+	- Protocol Buffers do not have a standard canonical encoding.
+		Unmarshalling and marshalling does not guarantee a binary match,
+		but signing algorithms operate on binary data.
+	- Protocol Buffers ignores data not listed in the schema by default,
+		but we should never sign unknown data.
+
+To solve these concerns we must sign and verify data encoded using Protocol Buffers carefully.
+
+Replication file contains a required header field, currently this must match "rtrade-replication v0".
+Before signing a file, all protocol buffers structures are checked against unrecognised fields.
+The binary data passed to the signing algorithm is defined as follows:
+
+	Where all numbers and length information are encoded as variable length integers (encoding/binary.PutUvarint).
+
+	Where all bytes and strings are encoded with length prefixed.
+
+	The bytes to be signed are: Topic | AuthorId (peer.ID binary encoding) | Version | Replication (encoded protocol buffers bytes)
+
+
+Notable missing features that may or may not be worked on
+
+Additional logic before accepting an activation, to protect against resource overuse in a public network.
+
+Conditions for requesting and accepting deactivation on a server.
+
+Management and removal of old versions from local database.
+
+Encrypted subscription. If there is a need to hide update meta data.
+
+gRPC over libp2p.
+
+Advanced server listing and ordering features.
+
+Graceful server deprecation. This add a third observable state "deprecated".
+It can not become active, but can be connected to and serves existing data.
+
+Erasure coding.
```

Comment:
I think I should convert this from godoc to markdown format.

Reply:
Yea that would probably be good